tests/unittests/Foundation/NSScannerTests.mm

```diff
@@ -0,0 +1,602 @@
+    EXPECT_EQ(-HUGE_VAL, result);
+    EXPECT_EQ(expectedScanLocation += 16, scanner.scanLocation);
+    EXPECT_FALSE(scanner.atEnd);
+
+    ASSERT_TRUE(performScan(scanner, sel, &result));
+    EXPECT_EQ(HUGE_VAL, result);
+    EXPECT_EQ(expectedScanLocation += 15, scanner.scanLocation);
+    EXPECT_TRUE(scanner.atEnd);
+}
+
+TEST(NSScanner, ScanHexDouble) {
+    testScanHexFloatingPoint<double>(@selector(scanHexDouble:));
+
+    // If the mantissa has too many bits to be represented, scanning should cut off the latter bits
+    NSScanner* scanner = [NSScanner scannerWithString:@"0x1.23456789aabbccddeeffp+3"];
+    double result;
+    ASSERT_TRUE([scanner scanHexDouble:&result]);
+    EXPECT_EQ(0x1.23456789aabbccp+3, result);
+    EXPECT_TRUE(scanner.atEnd);
+}
+
+TEST(NSScanner, ScanHexFloat) {
+    testScanHexFloatingPoint<float>(@selector(scanHexFloat:));
+
+    // If the mantissa has too many bits to be represented, scanning should cut off the latter bits
+    NSScanner* scanner = [NSScanner scannerWithString:@"0x1.23456789aabbccddeeffp+3"];
+    float result;
+    float expected = 0x1.234567p+3; // Declare as float here so it doesn't get instantiated as a double
+    ASSERT_TRUE([scanner scanHexFloat:&result]);
+    EXPECT_EQ(expected, result);
+    EXPECT_TRUE(scanner.atEnd);
+}
+
+template <typename TIntegral>
+static void testScanHexIntegral(SEL sel) {
+    NSScanner* scanner = [NSScanner scannerWithString:@":123 :DEADBEEF 0xDEADBEEF 0Xdeadbeef 0xEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE"];
+    scanner.charactersToBeSkipped = [NSCharacterSet characterSetWithCharactersInString:@": "];
+    NSUInteger expectedScanLocation = 0;
+    TIntegral result;
+
+    ASSERT_TRUE(performScan(scanner, sel, &result));
+    EXPECT_EQ(0x123, result);
+    EXPECT_EQ(expectedScanLocation += 4, scanner.scanLocation);
+    EXPECT_FALSE(scanner.atEnd);
+
+    ASSERT_TRUE(performScan(scanner, sel, &result));
+    EXPECT_EQ(0xDEADBEEF, result);
+    EXPECT_EQ(expectedScanLocation += 10, scanner.scanLocation);
+    EXPECT_FALSE(scanner.atEnd);
+
+    // Capitalization shouldn't matter
+    ASSERT_TRUE(performScan(scanner, sel, &result));
+    EXPECT_EQ(0xDEADBEEF, result);
+    EXPECT_EQ(expectedScanLocation += 11, scanner.scanLocation);
+    EXPECT_FALSE(scanner.atEnd);
+
+    ASSERT_TRUE(performScan(scanner, sel, &result));
+    EXPECT_EQ(0xDEADBEEF, result);
+    EXPECT_EQ(expectedScanLocation += 11, scanner.scanLocation);
+    EXPECT_FALSE(scanner.atEnd);
+
+    // Overflow should be valid and return the maximum representable value
+    ASSERT_TRUE(performScan(scanner, sel, &result));
+    EXPECT_EQ(std::numeric_limits<TIntegral>::max(), result);
+    EXPECT_EQ(expectedScanLocation += 40, scanner.scanLocation);
+    EXPECT_TRUE(scanner.atEnd);
+}
+
+TEST(NSScanner, ScanHexInt) {
+    testScanHexIntegral<unsigned int>(@selector(scanHexInt:));
+}
+
+TEST(NSScanner, ScanHexLongLong) {
+    testScanHexIntegral<unsigned long long>(@selector(scanHexLongLong:));
```

Comment:
I'd like to see an additional test for something that fits in 64 but not 32; this doesn't have that